File: python/final_nonfinal_words.py
from soniox.transcribe_live import transcribe_stream
from soniox.speech_service import SpeechClient


def iter_audio():
    with open("../test_data/test_audio_long.flac", "rb") as fh:
        while True:
            audio = fh.read(1024)
            if len(audio) == 0:
                break
            yield audio


def main():
    with SpeechClient() as client:
        # All final words will be collected in this list.
        all_final_words = []

        for result in transcribe_stream(iter_audio(), client):
            # Split current result response into final words and non-final words.
            final_words = []
            non_final_words = []
            for word in result.words:
                if word.is_final:
                    final_words.append(word.text)
                else:
                    non_final_words.append(word.text)

            # Append current final words to all final words.
            all_final_words += final_words

            # Print all final words and current non-final words.
            all_final_words_str = " ".join(all_final_words)
            non_final_words_str = " ".join(non_final_words)
            print(f"Final: {all_final_words_str}")
            print(f"Non-final: {non_final_words_str}")
            print("-----")


if __name__ == "__main__":
    main()

File: node/final_nonfinal_words.js
const fs = require("fs");
const { SpeechClient } = require("@soniox/soniox-node");

/**
 * Set your Soniox Cloud API key:
 * from command line: export SONIOX_API_KEY=<YOUR-API-KEY>
 * or
 * pass config object: new SpeechClient({ api_key: "<YOUR-API-KEY>" })
 */
const speechClient = new SpeechClient();

(async function () {
  // All final words will be collected in this list.
  let all_final_words = [];

  const onDataHandler = async (result) => {
    if (!result) {
      return;
    }

    const final_words = [];
    const non_final_words = [];

    for (word of result.words) {
      if (word.is_final) {
        final_words.push(word.text);
      } else {
        non_final_words.push(word.text);
      }
    }

    // Append current final words to all final words.
    all_final_words = all_final_words.concat(final_words);

    console.log(`Final: ${all_final_words.join(" ")}`);
    console.log(`Non-final: ${non_final_words.join(" ")}`);
    console.log("-----");
  };

  const onEndHandler = (error) => {
    if (error) {
      console.log(`Transcription error: ${error}`);
    }
  };

  // transcribeStream returns object with ".writeAsync()" and ".end()" methods
  // use them to send data and end stream when done.
  const stream = speechClient.transcribeStream(
    { include_nonfinal: true },
    onDataHandler,
    onEndHandler
  );

  const CHUNK_SIZE = 1024;
  const streamSource = fs.createReadStream(
    "../test_data/test_audio_long.flac",
    {
      highWaterMark: CHUNK_SIZE,
    }
  );

  // Simulate data streaming
  for await (const audioChunk of streamSource) {
    await stream.writeAsync(audioChunk);
  }

  stream.end();
})();
